### ISSUE TITLE
duplicity: Setting B2 credentials in `targetUrl` property without logging them to systemd

### ISSUE THREAD

**Issue by Ruin0x11** [NixOS/nixpkgs#127844]
**Describe the bug**
I'm rather new to nix, so please let me know if this isn't a valid issue.

I want to use `duplicity` to backup my server. It has a `targetUrl` property to specify the backend used. I use B2, and this is how they recommend building the URL:

```
b2://[keyID]:[application key]@[B2 bucket name]
```

I don't think these credentials can be specified in the environment like with the AWS backend; they have to be present in the URL itself. So I tried using shell expansion:

```nix
services.duplicity.targetUrl = "b2://$B2_KEYID:$B2_APPLICATION_KEY@$B2_BUCKET_NAME/backup";
services.duplicity.secretFile = "/etc/duplicity_secrets";
```

But this didn't work as I expected because the systemd script inside the `duplicity` package escapes the final shell command.

https://github.com/NixOS/nixpkgs/blob/465e5fc8cd73eca36e6bc4a320fdac8bd50bb160/nixos/modules/services/backup/duplicity.nix#L161-L166

I worked around this by grabbing the credentials from a `secret.nix` file and doing a string substitution.

```nix
services.duplicity.targetUrl = "b2://${secrets.duplicity.account}:${secrets.duplicity.key}@${secrets.duplicity.bucket}/backup";
```

However, this is less than ideal since the full command with the secrets present is logged to systemd's journal, because the generated script uses `set -x`. Is there at least some way to ensure this doesn't happen?

**To Reproduce**
As above.

**Expected behavior**
Using `duplicity` with the `b2://` backend without any secrets being logged.

**Notify maintainers**
@symphorien

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.43, NixOS, 21.05.1105.6613a30c5e3 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.12`
 - channels(root): `"nixos-21.05.1105.6613a30c5e3"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by symphorien** [comment]
I suspect that if you do not specify a password on the command line, the password is loaded from the `FTP_PASSWORD` environment variable (even for non ftp backends). The other environment variables are public, right?

**Comment by Ruin0x11** [comment]
You're right actually, sorry. For reference I used to have a script I would run via sudo that was unreadable by regular users, to prevent the credentials from being exposed. I will say that's a pretty terrible name, though...

But I'm not sure if the others should be public. They basically amount to your API client ID and the internal name of the storage bucket you choose. I would prefer not to have those exposed, but there's probably not much harm in doing so.

**Comment by uninsane** [comment]
right now i [patch](https://git.uninsane.org/colin/nix-files/src/commit/cb00ae4f9276d9b4acc1aad0c55ad0e6b2a92d75/nixpatches/07-duplicity-rich-url.patch) nixpkgs to have  a `duplicity.escapeUrl` option which if set to false allows $shell_variable access in the url.

i’m thinking better UX would be either a `duplicity.targetUrlExpr = “$MY_URL”` option or `duplicity.targetUrlFile = “/run/secrets/duplicity_url”`. i guess the latter is more robust against quoting errors, etc, but might require a more complicated wrapper.

if we can agree on what a good solution would look like i’m happy to put that together and send out a PR.

**Comment by symphorien** [comment]
even if the nixos module ends up executing this shell script: `duplicity verb $MY_URL` then $MY_URL gets leaked through /proc (any process can obtain it the same way htop does). So the only safe way to pass a password to duplicity is through the FTP_PASSWORD environment variable.

**Comment by uninsane** [comment]
> even if the nixos module ends up executing this shell script: `duplicity verb $MY_URL` then $MY_URL gets leaked through /proc (any process can obtain it the same way htop does). So the only safe way to pass a password to duplicity is through the FTP_PASSWORD environment variable.

good to point out. this is about more than just the password field (i.e. application key) though. i can't convince myself that having the key ID and bucket name be public is entirely benign. since i share my nix config publicly, those fields would be public to the world without some patching: narrowing that to just those with access to the machine is at least an improvement, if imperfect.

reading the manpage, i see no way to pass the key id or bucket name other than as CLI params -- except that _maybe_ one could (ab)use the `multi:///path/to/config.json` multi-backend protocol and thereby place the actual b2 URL on-disk in `config.json` where it should see less exposure.

<https://duplicity.gitlab.io/stable/duplicity.1.html#a-note-on-multi-backend>

**Comment by symphorien** [comment]
If you consider the rest of the URL secret, the limitation is then in duplicity, and I think the current module is not limiting. I'm inclined to close this issue, unless you disagree.
